### DISCUSSION TITLE
Now v18.0.0 environment variables vs secrets

### DISCUSSION THREAD

**Discussion by willemliu** [vercel/vercel#4065] [General] [answered]
I'm wondering what happens to Now Secrets now that Environment variables feature is available. Do we need to migrate our secrets to environment variables? Are secrets deprecated or will they be deprecated?

**Comment by Vadorequest** [comment, marked as answer]
@willemliu Both are kept.

- Secrets are global, shared across all projets
- Env vars are scoped by project

**Reply by styfle** [reply]
A few more differences here: https://twitter.com/styfle/status/1250043300557570048

1. now env is per project, secrets are global
2. now env creates a var, secrets must assign a var in `now.json`
3. now env reads stdin so you can pipe the value from a file
4. now env swaps env values upon merge to master
5. now env has the concept of "system" vars which gives you greater control over which values our system injects into your deployment

**Comment by ivan-kleshnin** [comment]
1) Why ENV variables are polluting the list of secrets?

```
  name                                created
  github-app-secret-preview-8xfy      3m ago
  node-env-preview-uti3               4m ago
  jwt-secret-preview-xo34             4m ago
  postgres-uri-preview-gfgd           4m ago
  github-app-secret-development-e6zj  1h ago
```

Even if they are implemented with the same underlying mechanics, shouldn't they be hidden?

2) If the (env-) secret is removed, the env is staying (empty). Such a big surface for bugs :(

3) The list of envs (and secrets as well) should be alphabetically sorted in UI and console. Now it's hard to compare env values between Environments :(

**Reply by willemliu** [reply]
This means that those secrets are global as well then...

**Reply by leo** [reply]
@ivan-kleshnin Thank you for sharing this feedback!

Think of Environment Variables as a way to attach Secrets to a project. That's why creating a Environment Variable nowadays creates a Secret for you – to provide you with the best security by default.

In the future, however, when adding Environment Variables, you'll be able to choose between:

1. Using a plaintext value
2. Using an existing global Secret
3. Creating a new global Secret

The only reason why we're creating Secrets automatically for you today, is so that you can get the highest security option straight out of the box.

**Reply by Vadorequest** [reply]
@leo Just curious about what would happen if I create a secret and an env var using the same name. Is it allowed? Does one take precedence over the other? Could you share an example? :)

**Reply by willemliu** [reply]
By the looks of it the env var created in the dashboard is suffixed with a hash.

**Reply by Vadorequest** [reply]
I wasn't sure if it had been written by the author "on purpose" to expose what had been created though secret/env, or if it's done by Zeit automatically. I'm still unsure.

**Reply by willemliu** [reply]
> I wasn't sure if it had been written by the author "on purpose" to expose what had been created though secret/env, or if it's done by Zeit automatically. I'm still unsure.

I can confirm that it's suffixed with some env variables in the dashboard I've setup in the dashboard. Named it the same as the ones I already had setup through `now secret add`.

**Reply by emregency** [reply]
> @ivan-kleshnin Thank you for sharing this feedback!
> 
> Think of Environment Variables as a way to attach Secrets to a project. That's why creating a Environment Variable nowadays creates a Secret for you – to provide you with the best security by default.
> 
> In the future, however, when adding Environment Variables, you'll be able to choose between:
> 
> 1. Using a plaintext value
> 2. Using an existing global Secret
> 3. Creating a new global Secret
> 
> The only reason why we're creating Secrets automatically for you today, is so that you can get the highest security option straight out of the box.

👍 I am so much looking for the new options to be able to select from? Any ETA for that? It's a burning topic for us right now. :)

Is there any way to differentiate the build.env _e.g. front-end_ as CRA from env _e.g. functions_ in the current Environment Variables UI?
